### PR TITLE
[choreo] Adds CLI support for generating trajectories

### LINKED ISSUE #36: Add commandline interface
In the spirit of one-click builds, I'd like the build task of a freshly cloned robot project to noninteractively generate trajectories and output them to the deploy folder for GradleRIO to deploy. A noninteractive commandline mode would facilitate this.

This makes tweaking waypoints before a match via the project JSON or via the UI less error-prone, because it ensures the generated outputs during a deploy are up to date.

### PR DESCRIPTION
**Pull request by oh-yes-0-fps**
Closes #36

Not a final implementation

If no args are provided it will launch GUI despite what help says

```
Usage: Choreo.exe [OPTIONS] --chor <path/to/myproject.chor>

Options:
  -g, --generate
  -h, --help      Print help

File Options:
      --chor <path/to/myproject.chor>
      --traj <test> //supports multiple inputs

Formating Options:
  -f, --format <(pretty|json|slim)>  [default: slim]
  ```
  
  right now you need a `--chor`, at least 1 `--traj` and `-g` for it to do anything

### CODE REVIEW

**Comment by oh-yes-0-fps** [comment]
Added the ability to open Choroe by doing `Choreo --chor path/to/myProject.chor`

I also overhauled the Tauri Commands to make them more isolated and reduce the reliance on absolute paths in the frontend.

More work still needs to be done on the frontend to make it follow this paradigm

File: src/components/field/svg/constraintDisplay/FieldConstraintAddLayer.tsx [hunk 1]
```diff
@@ -78,7 +79,7 @@ class FieldConstraintsAddLayer extends Component<Props, State> {
           showLines={false}
           id="add-first-circles"
           onCircleClick={(id) => {
-            console.log("constraint from: ", id);
+            TraceDebug("constraint from: ", id);
```
Comment (shueja) [inline src/components/field/svg/constraintDisplay/FieldConstraintAddLayer.tsx:82]: remove log

File: src/components/field/svg/constraintDisplay/FieldConstraintDisplayLayer.tsx [hunk 2]
```diff
@@ -19,9 +20,9 @@ type Props = {
   lineColor: string;
 };
 function FieldConstraintDisplayLayer(props: Props) {
-  console.log("FieldConstraintDisplayLayer", doc.isSidebarConstraintSelected);
```
Comment (shueja) [inline src/components/field/svg/constraintDisplay/FieldConstraintDisplayLayer.tsx:22]: remove both logs here

File: src/document/DocumentManager.ts [hunk 3]
```diff
@@ -715,14 +669,14 @@ export function select(item: SelectableItemTypes) {
...
-    doc.pathlist.paths.get(uuid)?.setName(newName);
-    const newPath = await getTrajFilePath(uuid);
-    if (oldPath !== null) {
-      Promise.all([Commands.deleteFile(oldPath[0], oldPath[1])])
-        .then(() => writeTrajectory(() => newPath, uuid))
+    const traj = doc.pathlist.paths.get(uuid);
+    if (traj) {
+      await Commands.renameTraj(traj.serialize, newName)
+        .then((new_traj) => {
+          doc.pathlist.addPath(newName, true, new_traj);
```
Comment (shueja) [inline src/document/DocumentManager.ts:676]: What does this do with the old path?

File: src/document/tauriTracing.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+/**
+ * A console.warn substitute that logs the warning to the native tracing pipeline.
+ *
+ * @param message The message to be logged. This will be converted to a string.
+ * @param optionalParams Additional parameters to be logged.
+ */
+export function TraceWarn(message: any, ...optionalParams: any[]): void {
+  console.log(message, optionalParams);
```
Comment (shueja) [inline src/document/tauriTracing.ts:84]: should be console.warn

File: src/document/tauriTracing.ts [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+/**
+ * A console.error substitute that logs the error to the native tracing pipeline.
+ *
+ * @param message The message to be logged. This will be converted to a string.
+ * @param optionalParams Additional parameters to be logged.
+ */
+export function TraceError(message: any, ...optionalParams: any[]): void {
+  console.log(message, optionalParams);
```
Comment (shueja) [inline src/document/tauriTracing.ts:121]: should be console.error

**Review by shueja — CHANGES_REQUESTED** [review]

File: src/document/DocumentManager.ts (hunk 3 above)
Comment (oh-yes-0-fps) [reply src/document/DocumentManager.ts:676]: Is just deleting the old one from the pathlist appropriate?

Comment (shueja) [reply src/document/DocumentManager.ts:677]: Actually I don’t see the benefit of deleting and tea-adding over just changing the name. For one thing, you won’t be able to keep the same waypoint selected through the delete and readd

**Comment by oh-yes-0-fps** [comment]
```bash
Choreo CLI

Usage: Choreo [OPTIONS]

Options:
  -h, --help     Print help
  -V, --version  Print version

File Options:
      --chor <path/to/myproject.chor>
      --traj <trajName1,trajName2>     Adds trajectories to be used by cli actions
      --all-traj                       The same as doing --traj for all trajectories in the project

Action Options:
  -g, --generate  Generate the provided trajectories for the project

Advanced Options:
  -f, --free       Free the process to run in another process and return immediately
      --ipc <IPC>  A serialized IPC handle to the parent process
```

https://github.com/user-attachments/assets/691a158b-e13f-4828-88b9-27d9af201e1a

**Review by shueja — APPROVED** [review]
I don't particularly care, but we might want to change the wording "solver killed" to "cancelled" especially since it shows up to the user as `Solver error: "Solver killed"`.

Also it's unclear whether the traj names passed to the cli are with or without absolute paths and/or file extensions.

Otherwise looks good to me. Conditionally approved on those two wording/doc cleanups

**Review by Gold856 — CHANGES_REQUESTED** [review]
Closing is bugged. Trying to close the Choreo window is impossible if you either didn't save the file or didn't open a file. If you did either of those, and then try to close Choreo, it will prompt you to save the project, but the file dialog prompts you to save a whole new file (as opposed to Choreo just exiting).

Also, this does not fix #611.

**Comment by oh-yes-0-fps** [comment]
> Some of the CLI options are broken too. `--help` and `--version` don't print out anything and Choreo just returns with exit code 0. `--chor` doesn't seem to accept relative paths.

were you running it in the vscode integrated terminal? all these issues are only replicable when using that. I tracked down and fixed the "not being able to close the app issue", it was added in my last commit so that's why I didn't notice.

**Comment by Gold856** [comment]
I was running Choreo in Windows Terminal with PowerShell. My command looks like this: `C:\Users\gold856\AppData\Local\Choreo\Choreo.exe --chor D:\Choreo\chor.chor --all-traj --generate`

**Comment by oh-yes-0-fps** [comment]
Ok fixed every issue you had i was able to replicate, the `--help` and `--version` stuff has always worked and is provided by a library so it would be hard to break.

**Comment by Gold856** [comment]
Can confirm that closing and #611 are fixed. The CLI args still don't work for me though.

**Comment by oh-yes-0-fps** [comment]
#624 and #597 are technically implemented in this too

**Comment by calcmogul** [comment]
This PR broke for me when generating a trajectory with two waypoints:
```
INFO | 21:23:33.048| src/tauri.rs:60                 | Starting Choreo Gui
TRACE| 21:23:33.049| unknown:-1                      | registering event source with poller: token=Token(1), interests=READABLE

** (choreo:2334484): WARNING **: 14:23:33.148: webkit_settings_set_enable_offline_web_application_cache is deprecated and does nothing.
TRACE| 21:23:34.309| unknown:-1                      | registering event source with poller: token=Token(1), interests=READABLE
DEBUG| 21:23:35.553| src-core/src/file_management/mod.rs:154| Setting deploy path to 
DEBUG| 21:24:11.578| DocumentManager.ts:509          | generateWithToastsAndExport 046f3e8c-e818-458e-a456-e826e3ecb7e2
DEBUG| 21:24:11.578| chunk-6OOF5XIV.js:1117          | generatePathWithToasts 046f3e8c-e818-458e-a456-e826e3ecb7e2
dt estimate: 0.1 - total time estimate: 1.540249126749439
ERROR| 21:24:11.605| react-toastify.js:114           | generatePathWithToasts: TypeError: undefined is not an object (evaluating 'caller.includes')
```

**Comment by oh-yes-0-fps** [comment]
> This PR broke for me when generating a trajectory with two waypoints in the GUI. I get the following error and no trajectory.

May be a linux only issue?

https://github.com/user-attachments/assets/ad3c5fc3-5b65-40c3-b831-986d6b18cca4

**Comment by oh-yes-0-fps** [comment]
The only thing left in this pr that's blatantly non-functional is frontend tracing locations on webkit

**Review by shueja — APPROVED** [review]

**Pull request merged**